code_generator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs

```diff
@@ -1,42 +1,171 @@
+using System.Collections.Generic;
 using Improbable.CodeGeneration.Model;
 using Improbable.CodeGeneration.Utils;
 
 namespace Improbable.Gdk.CodeGenerator
 {
     public class UnityFieldDetails
     {
+        public struct FieldTypeInfo
+        {
+            public TypeInfo? SingularTypeInfo;
+            public TypeInfo? ListTypeInfo;
+            public TypeInfo? OptionTypeInfo;
+            public KeyValuePair<TypeInfo, TypeInfo> MapTypeInfo;
+        }
+
+        public struct TypeInfo
+        {
+            public bool IsPrimitive;
+            public bool IsEnum;
+            public string Type;
+            public SerializationFunctionStrings SerializationFunctions;
+        }
+
+        // NOTE: These will be context dependent! If the type is primitive,
+        // these are called like SchemaObject.<function>(field_number). If
+        // the type is user defined, it will be called like:
+        // global::<type>.<function>(SchemaObject, field_number)
+        public struct SerializationFunctionStrings
+        {
+            public string Serialize;
+            public string Deserialize;
+            public string GetCount;
+        }
+
         public string Type;
+        public FieldTypeInfo SchemaTypeInfo;
         public string PascalCaseName;
         public string CamelCaseName;
         public FieldDefinitionRaw RawFieldDefinition;
+        public int FieldNumber;
 
-        public UnityFieldDetails(FieldDefinitionRaw rawFieldDefinition)
+        public UnityFieldDetails(FieldDefinitionRaw rawFieldDefinition, HashSet<string> enumSet)
         {
             PascalCaseName = Formatting.SnakeCaseToCapitalisedCamelCase(rawFieldDefinition.name);
             CamelCaseName = Formatting.SnakeCaseToCamelCase(rawFieldDefinition.name);
             var packagePrefix = UnityTypeMappings.PackagePrefix;
             RawFieldDefinition = rawFieldDefinition;
+            FieldNumber = rawFieldDefinition.Number;
 
             if (rawFieldDefinition.IsOption())
             {
                 var containedType = GetTypeFromTypeReference(rawFieldDefinition.optionType.valueType, packagePrefix);
```

Comment:
move `rawFieldDefinition.optionType.valueType` into a variable to avoid having to use it everywhere ;)